### PR TITLE
Multiarch support

### PR DESCRIPTION
**Pull request by gemagomez**
Making architecture a first class citizen on mr-provisioner:

- Modify models for machines, images to be arch specific
- Add a new file (image) type, `bootloader`
- A single architecture can have multiple sub-architectures. Bootloaders are specified on a sub-architecture basis.
- Allowing to serve a bootfile that is specific to the sub-architecture of a machine
- Adding UI bits to be able to manage arch/subarchs
- Migrating existing setups to have a default arch/subarch that can later be renamed
- Modify API to make use of arch/subarchs
- Modify/add tests for multiarch

Architectures would be things like x86_64, arm64, etc, whilst sub-architectures would be BIOS, EFI, etc. A machine and/or image is (tends to be?) architecture-specific, but it might support multiple boot mechanisms (aka subarchitectures).

### CODE REVIEW

File: mr_provisioner/api/v1/controllers.py [hunk 1]
```diff
@@ -116,6 +119,7 @@ def is_int(s):
 
 machine_schema = Schema({
     Optional('netboot_enabled'): bool,
+    Optional('subarch'): Or(None, And(str, lambda s: validators.length(s, min=0, max=256))),
```
Comment (rengolin) [inline mr_provisioner/api/v1/controllers.py]: Shouldn't this have arch, too? Line 49 above has both.

Or is arch contained in subarch?

File: mr_provisioner/models.py [hunk 2]
```diff
@@ -403,6 +401,68 @@ def query_by_criteria(query_str, *, no_assignees=False):
...
+    @property
+    def machines(self):
+        return Machine.query.filter((Machine.arch_id == self.id)).all()
+
+    @staticmethod
+    def can_create(user):
+        return True if user.admin else False
+
+
+# Sub architectures are UEFI/BIOS/etc
```
Comment (rengolin) [inline mr_provisioner/models.py]: I assumed subarch was like ARMv8.x, AMD64 vs i386, etc.
I would call this bootarch, not sub arch

File: mr_provisioner/api/v1/controllers.py (hunk 1 above)
Comment (bwalex) [reply mr_provisioner/api/v1/controllers.py]: No. the PUT endpoint for a machine only allows changing things related to provisioning. It mostly maps to what happens when you click on "edit" near the provisioning bits & pieces in the UI.

File: mr_provisioner/admin/controllers.py [hunk 3]
```diff
@@ -249,6 +250,309 @@ class MachineType(graphene.ObjectType):
...
+
+        return (len(errors) == 0, errors)
+
+    @staticmethod
+    def mutate(root, args, context, info):
+        errors = []
+
+        if not Arch.can_create(g.user):
+            errors = ['Permission denied']
+            return CreateArch(arch=None, ok=False, errors=errors)
```
Comment (niedbalski) [inline mr_provisioner/admin/controllers.py]: Minor nitpick, I think is a pattern repeated over the code. What if instead of doing a return CreateArch on most of the cases, you just do a single return at the last part of the method?
return CreateArch(arch=arch, ok=len(errors) == 0, errors=errors)

File: mr_provisioner/admin/controllers.py [hunk 4]
```diff
@@ -249,6 +250,309 @@ class MachineType(graphene.ObjectType):
...
+            arch.description = args.get('description')
+
+            try:
+                db.session.commit()
+                db.session.refresh(arch)
+            except IntegrityError as e:
+                db.session.rollback()
+                return ChangeArch(arch=None, ok=False, errors=['An architecture with that name already exists'])
+
+            return ChangeArch(arch=arch, ok=True, errors=errors)
```
Comment (niedbalski) [inline mr_provisioner/admin/controllers.py]: Same nitpick as before

File: mr_provisioner/admin/controllers.py [hunk 5]
```diff
@@ -249,6 +250,309 @@ class MachineType(graphene.ObjectType):
...
+
+    ok = graphene.Boolean()
+    errors = graphene.List(graphene.String)
+    arch = graphene.Field(ArchType)
+
+    @staticmethod
+    def validate(args):
+        errors = []
+
+        return (len(errors) == 0, errors)
```
Comment (niedbalski) [inline mr_provisioner/admin/controllers.py]: Is this needed?

File: mr_provisioner/admin/controllers.py [hunk 6]
```diff
@@ -249,6 +250,309 @@ class MachineType(graphene.ObjectType):
...
+
+    @staticmethod
+    def validate(args, bootloader, arch):
+        errors = []
+
+        if not validators.length(args.get('name'), min=2, max=256):
+            errors.append('Architecture name must be between 2 and 256 characters long')
+
+        if not validators.length(args.get('description', ''), min=0, max=256):
+            errors.append('description must be between 0 and 256 characters long')
```
Comment (niedbalski) [inline mr_provisioner/admin/controllers.py]: /s/d/D/g/

File: mr_provisioner/admin/controllers.py [hunk 7]
```diff
@@ -249,6 +250,309 @@ class MachineType(graphene.ObjectType):
...
+
+    @staticmethod
+    def validate(args):
+        errors = []
+
+        if not validators.length(args.get('name'), min=2, max=256):
+            errors.append('Architecture name must be between 2 and 256 characters long')
+
+        if not validators.length(args.get('description', ''), min=0, max=256):
+            errors.append('description must be between 0 and 256 characters long')
```
Comment (niedbalski) [inline mr_provisioner/admin/controllers.py]: s/d/D/g

**Comment by niedbalski** [comment]
@gemagomez 

I've made a couple of comments on the code (nothing critical), and I have a couple of questions for you.
- Should be possible to create a sub-architecture without a bootloader associated to it? The UI allowed
me to perform that operation by assigning it to None and save. I think that the None value comes from the fact that I didn't have any pre-existing bootloader defined, but I do have a default_bootfile configured instead.  I think this validation isn't working properly.
image

- Also, when enabling netboot on a machine that has this arch -> subarch (and consequently this none bootloader) , I get the following message:
image
However, at the XHR side, I can see the validation being performed correctly:
```json
{"data":{"changeMachineNetboot":{"ok":false,"errors":["Cannot enable netboot without a subarchitecture with bootloader"],"machine":null,"__typename":"ChangeMachineNetboot"}}}
```

File: mr_provisioner/admin/controllers.py (hunk 3 above)
Comment (bwalex) [reply mr_provisioner/admin/controllers.py]: Well, early returns are good practice. Otherwise, you'd end up with way more nesting, which just makes things unreadable and hard to follow.

**Comment by bwalex** [comment]
@niedbalski 

Being able to create a sub-architecture without a bootloader is a useful thing to do, in that it allows you to think about architectures/subarchitecture that you will deal with before, necessarily, thinking about which exact bootloader binary you'll want.

@gemagomez: The second thing is clearly a bug - lines 977 and 980 in controllers.py (in `ChangeMachineNetboot.mutate`) should return machine=machine, not machine=None, since the machine has already been resolved at that point.

File: mr_provisioner/admin/controllers.py (hunk 5 above)
Comment (gemagomez) [reply mr_provisioner/admin/controllers.py]: It is there to make it consistent with everything else, simplifying, though.

File: mr_provisioner/admin/controllers.py (hunk 7 above)
Comment (gemagomez) [reply mr_provisioner/admin/controllers.py]: Done.

File: mr_provisioner/admin/controllers.py (hunk 6 above)
Comment (gemagomez) [reply mr_provisioner/admin/controllers.py]: Done.
